File: graph/multi-source-01-bfs/solution.py
# Stepsort · Multi-Source 0-1 BFS
# Category: Graph
# Animated walkthrough: https://stepsort.prakashraj.me/algorithm/multi-source-01-bfs

from collections import deque

def multi_source_01_bfs(graph, sources):
    dist = [float("inf")] * len(graph)
    dq = deque()
    for s in sources:
        dist[s] = 0
        dq.appendleft(s)
    while dq:
        u = dq.popleft()
        for v, w in graph[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                if w == 0:
                    dq.appendleft(v)
                else:
                    dq.append(v)
    return dist

graph = {
    0: [(1, 0), (2, 1)],
    1: [(3, 1)],
    2: [(3, 0), (4, 1)],
    3: [(5, 0)],
    4: [(5, 1)],
    5: [],
}
print(multi_source_01_bfs(graph, [0, 2]))

File: graph/multi-source-01-bfs/solution.cpp
// sortsort · Multi-Source 0-1 BFS
// Category: Graph
// Animated walkthrough: https://stepsort.prakashraj.me/algorithm/multi-source-01-bfs

vector<int> multiSource01BFS(vector<vector<pair<int,int>>>& graph, vector<int>& sources) {
    int n = graph.size();
    vector<int> dist(n, INT_MAX);
    deque<int> dq;
    for (int s : sources) { dist[s] = 0; dq.push_front(s); }
    while (!dq.empty()) {
        int u = dq.front(); dq.pop_front();
        for (auto [v, w] : graph[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                if (w == 0) dq.push_front(v);
                else dq.push_back(v);
            }
        }
    }
    return dist;
}
